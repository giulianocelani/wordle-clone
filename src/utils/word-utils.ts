import words from '../5-letters.json';

const MAX_WORD_LENGTH = 5;

function getRandomWord() {
	return words[Math.floor(Math.random() * words.length)];
}

export { getRandomWord, MAX_WORD_LENGTH };
