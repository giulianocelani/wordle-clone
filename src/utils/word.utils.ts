import words from '../5-letters.json';

const MAX_WORD_LENGTH = 5;

enum LetterState {
	MISS = 'MISS',
	MATCH = 'MATCH',
	PRESENT = 'PRESENT'
}

function getRandomWord() {
	return words[Math.floor(Math.random() * words.length)].toUpperCase();
}

function evaluateGuess(answer: string, guess: string): LetterState[] {
	if (answer.length !== guess.length && guess.length !== MAX_WORD_LENGTH) {
		return [];
	}

	const result: LetterState[] = Array(MAX_WORD_LENGTH).fill(LetterState.MISS);

	const lettersToCheck = answer.split('');
	const guessLetters = guess.split('');

	guessLetters.forEach((letter, i) => {
		if (answer[i] === letter) {
			result[i] = LetterState.MATCH;
			lettersToCheck.splice(i === lettersToCheck.length ? i - 1 : i, 1);
		}
	});

	guessLetters.forEach((letter, i) => {
		if (lettersToCheck.includes(letter) && result[i] !== LetterState.MATCH) {
			result[i] = LetterState.PRESENT;
			lettersToCheck.splice(lettersToCheck.indexOf(letter), 1);
		}
	});

	return result;
}

function isValidWord(word: string): boolean {
	return words.includes(word.toLowerCase());
}

export {
	evaluateGuess,
	getRandomWord,
	isValidWord,
	LetterState,
	MAX_WORD_LENGTH
};
