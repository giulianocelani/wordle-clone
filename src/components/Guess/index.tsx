import { IGuess } from '../../store';
import { MAX_WORD_LENGTH } from '../../utils';
import Letter from '../Letter';

type IProps = {
	guess: IGuess;
	isInvalid: boolean;
};

const Guess = ({ guess, isInvalid }: IProps) => {
	const letters = guess.word ? guess.word.split('') : [];

	const letterList = [
		...letters,
		...Array(MAX_WORD_LENGTH - letters.length).fill('')
	];

	return (
		<div
			className={`w-full grid grid-cols-5 gap-2 ${
				isInvalid ? 'animate-shake' : ''
			}`}
		>
			{letterList.map((letter, index) => (
				<Letter
					key={`letter-${index}`}
					letter={letter}
					state={guess.state && guess.state[index]}
				/>
			))}
		</div>
	);
};

export default Guess;
