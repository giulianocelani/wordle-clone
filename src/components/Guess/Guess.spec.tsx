import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { MAX_WORD_LENGTH } from '../../utils';

import Guess from '.';

describe('Guess', () => {
	const wrapper = mount(<Guess guess={{ word: 'HELLO' }} isInvalid={false} />);

	it('default renders correctly', () => {
		expect(toJson(wrapper, { mode: 'deep' })).toMatchSnapshot();
	});

	it('renders only maximum number of letters', () => {
		expect(
			wrapper.find(`div.grid-cols-${MAX_WORD_LENGTH}`).children().length
		).toBe(MAX_WORD_LENGTH);
	});

	it('renders letters in the correct order', () => {
		expect(
			wrapper
				.find(`div.grid-cols-${MAX_WORD_LENGTH}`)
				.children()
				.map((child) => child.props().letter)
		).toEqual(['H', 'E', 'L', 'L', 'O']);
	});
});
