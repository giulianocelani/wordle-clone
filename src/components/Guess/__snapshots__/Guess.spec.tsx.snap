// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Guess default renders correctly 1`] = `
<div
  className="grid grid-cols-5 gap-4 "
>
  <div
    className="border rounded-md flex items-center justify-center w-16 h-16 "
  >
    <h2
      className="font-semibold text-3xl "
    >
      H
    </h2>
  </div>
  <div
    className="border rounded-md flex items-center justify-center w-16 h-16 "
  >
    <h2
      className="font-semibold text-3xl "
    >
      E
    </h2>
  </div>
  <div
    className="border rounded-md flex items-center justify-center w-16 h-16 "
  >
    <h2
      className="font-semibold text-3xl "
    >
      L
    </h2>
  </div>
  <div
    className="border rounded-md flex items-center justify-center w-16 h-16 "
  >
    <h2
      className="font-semibold text-3xl "
    >
      L
    </h2>
  </div>
  <div
    className="border rounded-md flex items-center justify-center w-16 h-16 "
  >
    <h2
      className="font-semibold text-3xl "
    >
      O
    </h2>
  </div>
</div>
`;
